File: ratio_backend/domain/assessment.py
"""Assessment domain types for Ratio scoring, rationale, and publishing."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, datetime
from enum import Enum


class PublicationStatus(str, Enum):
    """Lifecycle states for an assessment revision."""

    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


@dataclass(slots=True)
class FactorScores:
    """The four methodology factor scores constrained to a 0-10 range."""

    debt: int
    market_share_change: int
    market_definition_change: int
    relative_valuation: int

    def as_dict(self) -> dict[str, int]:
        """Serialize scores for persistence or APIs."""
        return {
            "debt": self.debt,
            "market_share_change": self.market_share_change,
            "market_definition_change": self.market_definition_change,
            "relative_valuation": self.relative_valuation,
        }

    def validate(self, *, min_score: int = 0, max_score: int = 10) -> None:
        """Ensure each score stays within the allowed range."""
        for name, value in self.as_dict().items():
            if value < min_score or value > max_score:
                raise ValueError(f"{name} must be between {min_score} and {max_score}, got {value}")


@dataclass(slots=True)
class FactorAssessment:
    """A single factor score and its associated internal or public rationale."""

    factor_key: str
    factor_label: str
    score: int
    score_min: int = 0
    score_max: int = 10
    internal_rationale: str | None = None
    public_rationale_override: str | None = None
    sort_order: int = 0


@dataclass(slots=True)
class SizingResult:
    """Calculated portfolio sizing output derived from factor scores."""

    total_score: int
    normalized_score: float
    custom_beta: float
    base_position_size: float
    suggested_position_size: float

    def as_dict(self) -> dict[str, float | int]:
        """Serialize sizing data for persistence or APIs."""
        return {
            "total_score": self.total_score,
            "normalized_score": self.normalized_score,
            "custom_beta": self.custom_beta,
            "base_position_size": self.base_position_size,
            "suggested_position_size": self.suggested_position_size,
        }


@dataclass(slots=True)
class AssessmentEvidenceLink:
    """A link between an assessment or factor and a source document."""

    source_document_id: int
    factor_key: str | None = None
    relevance_rank: int = 0
    evidence_note: str | None = None
    used_by_llm: bool = False
    used_in_final_review: bool = False
    id: int | None = None


@dataclass(slots=True)
class LLMDraftMetadata:
    """Metadata captured when an LLM produces a first-pass assessment draft."""

    provider: str | None = None
    model: str | None = None
    prompt_version: str | None = None
    assumptions: list[str] = field(default_factory=list)
    raw_response: str | None = None


@dataclass(slots=True)
class AssessmentRecord:
    """A persisted assessment revision for one company."""

    company_id: int
    factor_scores: FactorScores
    sizing: SizingResult
    aggregate_score: int
    relative_score: float
    beta_like_score: float
    status: PublicationStatus = PublicationStatus.DRAFT
    calculation_version: str = "beta-v1"
    created_by: str | None = None
    updated_by: str | None = None
    published_by: str | None = None
    public_comment: str | None = None
    public_comment_enabled: bool = False
    internal_notes: str | None = None
    llm_metadata: LLMDraftMetadata | None = None
    as_of_date: date | None = None
    published_at: datetime | None = None
    factors: list[FactorAssessment] = field(default_factory=list)
    evidence_links: list[AssessmentEvidenceLink] = field(default_factory=list)
    id: int | None = None

File: crates/ratio-domain/src/assessment.rs
//! Assessment, factor, and sizing entities for the Ratio methodology.

use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PublicationStatus {
    Draft,
    Published,
    Archived,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FactorKey {
    Debt,
    MarketShareChange,
    MarketDefinitionChange,
    RelativeValuation,
}

impl FactorKey {
    pub fn label(self) -> &'static str {
        match self {
            Self::Debt => "Debt",
            Self::MarketShareChange => "Change in Market Share",
            Self::MarketDefinitionChange => "Change in Definition of Market",
            Self::RelativeValuation => "Relative Valuation",
        }
    }
}

#[derive(Debug, Error)]
pub enum DomainError {
    #[error("{field} must be between {min} and {max}, got {value}")]
    InvalidScore {
        field: &'static str,
        min: i32,
        max: i32,
        value: i32,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FourFactorScores {
    pub debt: i32,
    pub market_share_change: i32,
    pub market_definition_change: i32,
    pub relative_valuation: i32,
}

impl FourFactorScores {
    pub fn validate(&self, min: i32, max: i32) -> Result<(), DomainError> {
        for (field, value) in [
            ("debt", self.debt),
            ("market_share_change", self.market_share_change),
            ("market_definition_change", self.market_definition_change),
            ("relative_valuation", self.relative_valuation),
        ] {
            if value < min || value > max {
                return Err(DomainError::InvalidScore {
                    field,
                    min,
                    max,
                    value,
                });
            }
        }
        Ok(())
    }

    pub fn total(&self) -> i32 {
        self.debt
            + self.market_share_change
            + self.market_definition_change
            + self.relative_valuation
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FactorScore {
    pub factor_key: FactorKey,
    pub factor_label: String,
    pub score: i32,
    pub score_min: i32,
    pub score_max: i32,
    pub internal_rationale: Option<String>,
    pub public_rationale_override: Option<String>,
    pub sort_order: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssessmentEvidenceLink {
    pub id: Option<i64>,
    pub source_document_id: i64,
    pub factor_key: Option<FactorKey>,
    pub relevance_rank: i32,
    pub evidence_note: Option<String>,
    pub used_by_llm: bool,
    pub used_in_final_review: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SizingSnapshot {
    pub aggregate_score: i32,
    pub relative_score: f64,
    pub beta_like_score: f64,
    pub base_position_size: f64,
    pub suggested_position_size: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LlmDraftMetadata {
    pub provider: Option<String>,
    pub model: Option<String>,
    pub prompt_version: Option<String>,
    pub assumptions: Vec<String>,
    pub confidence: Option<String>,
    pub raw_response: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Assessment {
    pub id: Option<i64>,
    pub company_id: i64,
    pub version_number: i32,
    pub status: PublicationStatus,
    pub created_by: Option<String>,
    pub updated_by: Option<String>,
    pub published_by: Option<String>,
    pub as_of_date: Option<NaiveDate>,
    pub public_comment: Option<String>,
    pub is_public_comment_enabled: bool,
    pub internal_notes: Option<String>,
    pub calculation_version: String,
    pub sizing: SizingSnapshot,
    pub factors: Vec<FactorScore>,
    pub evidence_links: Vec<AssessmentEvidenceLink>,
    pub llm_metadata: Option<LlmDraftMetadata>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub published_at: Option<DateTime<Utc>>,
}
